File: python/languagecheck/translate.py
#	Copyright 2015, Google, Inc.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import urllib
import urllib2
import json
import logging
from google.appengine.api import memcache


class Translate():
	default_target = "en"
	baseurl = "https://www.googleapis.com/language/translate/v2"

	def __init__(self, key, target):
	    self.key = key
	    self.target = target
	    if len(target) == 0:
	    	self.target = default_target

	# Gets a languge identifier for a given string. Will try cache first.   
	def detect(self, key):

		data = memcache.get(key)

		if data is not None:
			logging.info("Cache hit")
		else:
			logging.info("Cache miss")
			data = self.detectFromAPI(str)
			memcache.set(key, data)

		label = self.getLanguageLabel(data)
		return {'data': data, 'label': label}


	# Sends a string to Google API and gets the language value for it. 
	def detectFromAPI(self, str):
		path = "/detect"

		str = str.encode("utf-8")


		params = urllib.urlencode({'q':str,'key':self.key})
		query = "?" + params
		url = self.baseurl + path + query;

		try:
			result = urllib2.urlopen(url)
			j = json.loads(result.read())
			return j['data']['detections'][0][0]['language']
		except urllib2.URLError, e:
			# You probably want to handle this error a little more completely. 
			print("Error:")
			print(e.reason)
			return "Error"

	# Gets a list of languages from Google API.
	def languagesFromAPI(self):
		path = "/languages"
		query = "?target=" + self.target + "&key=" + self.key
		url = self.baseurl + path + query;
		langs = {}

		try:
			result = urllib2.urlopen(url)
			j = json.loads(result.read())

			for item in j['data']['languages']:
				langs[ item['language']] = item['name']

			return langs
		except urllib2.URLError, e:
			# You probably want to handle this error a little more completely. 
			print("Error:")
			print(e.reason)
			return {"Error"}

	# Gets the label for a give language (lang) in target language (self) 
	# so getLanguageLabel("fr", "en") should yield "Anglais"
	#  Will try cache first.
	def getLanguageLabel(self, lang):
		key = "lang:" + self.target + ":" + lang

		data = memcache.get(key)

		if data is not None:
			logging.info("Cache hit")
			return data
		else:
			logging.info("Cache miss")
			langs = self.languagesFromAPI()

			for i in langs:
				tempkey = "lang:" + self.target +":" + i;
				memcache.set(tempkey, langs[i])

			data = langs[lang]
			return data

File: php/languagecheck/translate.php
<!--
	Copyright 2015, Google, Inc.
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
-->
<?php
/**
  * 
  * @package Translate
  * Encapsulates all the calls to Google Translation API
  */ 

class Translate { 
	var $baseurl = "https://www.googleapis.com/language/translate/v2";
	var $key = "";
	var $default_target = "";

	function __construct($key, $default_target="en") {
		$this->key = $key;
		$this->default_target = $default_target;
	}

	/**
	 * Calls Translate api to return a best guess at what language $str is written in.  
	 * @param string $str 
	 * @return string
	 */
	function detectFromAPI($str){
		$path = "/detect";
		$query = "?q=" . urlencode($str) . "&key=" . $this->key;
		$u = $this->baseurl . $path . $query;
		$response = file_get_contents($u, false);
		$result = json_decode($response);
		return $result->data->detections[0][0]->language;
	}

	/**
	 * Calls Translate api to return a hash of readable language labels keyed to language codes.   
	 * @return []
	 */
	function languagesFromAPI(){
		$path = "/languages";
		$query = "?target=" . $this->default_target . "&key=" . $this->key;
		$u = $this->baseurl . $path . $query;
		$response = file_get_contents($u, false);
		$result = json_decode($response);

		$langhash = [];

		foreach ($result->data->languages as &$value) {
		    $langhash[$value->language] = $value->name;
		}

		return $langhash;
	}
	
	/**
	 * Gets a language label from a language code. 
	 * Tries cache first, if not there, it populates cache. 
	 * @return string
	 */
	function getLanguageLabel($lang) {
		$key = "lang:" . $this->default_target .":" . $lang;
		$memcache = new Memcache;
		$data = $memcache->get($key);


		if ($data === false) {
			$hash = $this->languagesFromAPI();

			foreach ($hash as $key => $value) {
				$key = "lang:" . $this->default_target .":" . $key;
			    $memcache->set($key, $value);
			}

			$data = $hash[$lang];
		} 
		return $data;
	}

	/**
	 * Given a string returns the detected language and language code.
	 * Tries cache first, if not there, it populates cache.  
	 * @param string $str 
	 * @return obj
	 */
	function detect($str) {
		$key = $str;
		$memcache = new Memcache;
		$data = $memcache->get($key);

		if ($data === false) {
			$data = $this->detectFromAPI($str);
			$memcache->set($key, $data);
		}

		$label = $this->getLanguageLabel($data);
		$result = new stdClass(); 
		$result->label = $label;
		$result->data = $data;
		

		return json_encode($result);
	}

}

?>